File: services/auth-service-py/src/domain/use_cases/crear_usuario_use_case.py
from dataclasses import dataclass
from typing import Optional
import bcrypt

from ..repositories.i_usuario_repository import IUsuarioRepository
from ..entities.usuario import Usuario


@dataclass
class CrearUsuarioInput:
    nombre_completo: str
    email: str
    password: str
    rol_id: int
    sede_id: int
    zona_asignada: Optional[str] = None


class ConflictError(Exception):
    pass


class CrearUsuarioUseCase:
    def __init__(self, usuario_repo: IUsuarioRepository):
        self._repo = usuario_repo

    async def execute(self, input: CrearUsuarioInput) -> Usuario:
        existente = await self._repo.find_by_email(input.email)
        if existente:
            raise ConflictError(f"El email '{input.email}' ya está en uso")

        password_hash = bcrypt.hashpw(input.password.encode(), bcrypt.gensalt()).decode()

        usuario = await self._repo.create(
            nombre_completo=input.nombre_completo,
            email=input.email,
            password_hash=password_hash,
            rol_id=input.rol_id,
            sede_id=input.sede_id,
            zona_asignada=input.zona_asignada,
        )
        return usuario

File: services/auth-service/src/domain/use-cases/CrearUsuarioUseCase.ts
import { Rol, hashPassword, ConflictError, ForbiddenError, NotFoundError } from '@sec/shared';
import { IUsuarioRepository, CrearUsuarioDTO } from '../repositories/IUsuarioRepository';
import { ISedeRepository } from '../repositories/ISedeRepository';
import { Usuario } from '../entities/Usuario';

interface CrearUsuarioInput {
  nombreCompleto: string;
  email:          string;
  password:       string;
  rol:            Rol;
  sedeId:         number;
  zonaAsignada?:  string;
  creadoPor:      Usuario; // El usuario que ejecuta la acción
}

/**
 * Caso de uso: Crear Usuario
 * Valida permisos, verifica que la sede exista y que el email no esté en uso.
 */
export class CrearUsuarioUseCase {
  constructor(
    private readonly usuarioRepo: IUsuarioRepository,
    private readonly sedeRepo:   ISedeRepository,
  ) {}

  async execute(input: CrearUsuarioInput): Promise<Usuario> {
    const { nombreCompleto, email, password, rol, sedeId, zonaAsignada, creadoPor } = input;

    // Verificar permisos del creador
    if (!creadoPor.puedeCrearRol(rol)) {
      throw new ForbiddenError(`No puede crear usuarios con rol ${rol}`);
    }

    if (!creadoPor.puedeGestionarSede(sedeId)) {
      throw new ForbiddenError('No puede crear usuarios en otra sede');
    }

    // Verificar que la sede existe
    const sede = await this.sedeRepo.findById(sedeId);
    if (!sede) {
      throw new NotFoundError('Sede', sedeId);
    }

    // Verificar email único
    const emailExiste = await this.usuarioRepo.existeEmail(email);
    if (emailExiste) {
      throw new ConflictError(`El email '${email}' ya está registrado`);
    }

    const passwordHash = await hashPassword(password);

    const dto: CrearUsuarioDTO = {
      nombreCompleto,
      email,
      passwordHash,
      rol,
      sedeId,
      zonaAsignada,
    };

    return this.usuarioRepo.create(dto);
  }
}
